src/puppetlabs/puppetdb/cli/benchmark.clj

```diff
@@ -117,66 +113,67 @@
                            (rand-nth [1 2 3 4])))
   ([depth children]
-     (if (zero? depth)
-       {(random-string 10) (random-fact-value)}
-       {(random-string 10) (zipmap (repeatedly children #(random-string 10))
-                                   (repeatedly children #(random-structured-fact (rand-nth (range depth))
-                                                                                 (rand-nth (range children)))))})))
+   {(random-string 10)
+    (if (zero? depth)
+      (random-fact-value)
+      (zipmap (repeatedly children #(random-string 10))
+              (repeatedly children #(random-structured-fact (rand-nth (range depth))
+                                                            (rand-nth (range children))))))}))
 
 (defn maybe-tweak-catalog
   "Slightly tweak the given catalog, returning a new catalog, `rand-percentage`
   percent of the time."
-  [catalog rand-percentage]
+  [rand-percentage catalog]
   (if (< (rand 100) rand-percentage)
     (rand-catalog-mutation catalog)
     catalog))
 
-(defn update-catalog [catalog]
-  (assoc catalog "producer_timestamp" (time/now)))
+(defn update-producer-timestamp
+  "Updates the producer_timestamp to be current."
+  [entity]
+  (assoc entity "producer_timestamp" (time/now)))
 
-(defn update-report-run-fields
+(defn update-report
   "configuration_version, start_time and end_time should always change
    on subsequent report submittions, this changes those fields to avoid
    computing the same hash again (causing constraint errors in the DB)"
   [report]
-  (assoc report
-         "configuration_version" (ks/uuid)
-         "start_time" (time/now)
-         "end_time" (time/now)
-         "producer_timestamp" (time/now)))
+  (-> report
+      (assoc "configuration_version" (kitchensink/uuid)
+             "start_time" (time/now)
+             "end_time" (time/now))
+      update-producer-timestamp))
 
 (defn randomize-map-leaf
-  "Randomizes a fact leaf based on a percentage provided with `rp`."
-  [rp leaf]
-  (if (< (rand 100) rp)
-    (cond
-     (string? leaf) (random-string (inc (rand-int 100)))
-     (integer? leaf) (rand-int 100000)
-     (float? leaf) (* (rand) (rand-int 100000))
-     (ks/boolean? leaf) (random-bool))
-    leaf))
+  "Randomizes a fact leaf/"
+  [leaf]
+  (cond
+    (string? leaf) (random-string (inc (rand-int 100)))
+    (integer? leaf) (rand-int 100000)
+    (float? leaf) (* (rand) (rand-int 100000))
+    (kitchensink/boolean? leaf) (random-bool)))
 
 (defn randomize-map-leaves
-  "Runs through a map and randomizes leafs based on a percentage provided with
-   `rp`."
-  [rp value]
-  (cond
-   (map? value)
-   (ks/mapvals (partial randomize-map-leaves rp) value)
+  "Runs through a map and randomizes and random percentage of leaves."
+  [value & [rand-percentage]]
+  (let [rand-perc (or rand-percentage (rand-int 100))]
+    (cond
+      (map? value)
+      (kitchensink/mapvals (partial randomize-map-leaves rand-perc) value)
```

Comment:
oh I see, misunderstood your last comment.

We've never really nailed down what we mean by rand-perc applied to factsets (or what a "fact duplication rate" is), and we don't have good dashboard metrics to quantify the effect of the parameter yet. The existing doc string does say "updates the leaves" though.  For catalogs the update is all or nothing. I think the way we're doing it now seems more realistic, because having identical catalogs across the infrastructure seems more likely than having identical factsets, or factsets that only differ in certname.

I haven't looked to hard at it, but I don't see your "rp of rp" scenario in the existing code in master. Seems like rand-perc only comes into play on the outermost leaves (in randomize-map-leaf.) Maybe I'm missing something there though.